app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -168,39 +172,119 @@ public function indexAction()
+            $regionModel = Mage::getModel('directory/region');
+            $regionObj= $regionModel->loadByCode($address_data['region'], $address_data['country_id']);
+            $address_data['region_id'] = ($regionObj) ? $regionObj->getId() : null;
+
+            ////////////////////////////////////////////////////////////////////////
+            // Clear previously set estimates.  This helps if this
+            // process fails or is aborted, which will force the actual index action
+            // to get fresh data instead of reading from the session cache
+            ////////////////////////////////////////////////////////////////////////
+            $cacheKeyLocation = $quote->getId() . '_' . $address_data['country_id'];
+            $prefetchCacheKey = $this->generateCacheKey($quote, $address_data);
+            Mage::app()->getCache()->remove('quote_location_' . $cacheKeyLocation);
+            Mage::app()->getCache()->remove($prefetchCacheKey);
+            ////////////////////////////////////////////////////////////////////////
+
+            $quote->getShippingAddress()->addData($address_data);
+            $quote->getBillingAddress()->addData($address_data);
+            try {
+                $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
+            } catch (Exception $e) {
+                $estimate_response = null;
+            }
+
+            $lifeTime = 600; //10 minutes
+            Mage::app()->getCache()->save(
+                serialize($address_data),
+                'quote_location_' . $cacheKeyLocation,
+                array('BOLT_QUOTE_LOCATION'),
+                $lifeTime
+            );
+
+            Mage::app()->getCache()->save(
+                serialize($estimate_response),
+                $prefetchCacheKey,
+                array('BOLT_PREFETCH_SHIPPING_AND_TAX'),
+                $lifeTime
+            );
+        } else {
+            $address_data = array();
+        }
+
+        $response = Mage::helper('core')->jsonEncode(array('address_data' => $address_data));
+        $this->getResponse()->setHeader('Content-type', 'application/json');
+        $this->getResponse()->setHeader('X-Bolt-Cache-Hit', 'HIT');
+        $this->getResponse()->setBody($response);
+    }
+
+    /**
+     * @param        $quote Mage_Sales_Model_Quote
+     * @param        $addressData
+     * @return string
+     */
+    public function generateCacheKey($quote, $addressData)
+    {
+        /* @var $shippingAddress Mage_Sales_Model_Quote_Address */
+        $shippingAddress = $quote->getShippingAddress();
+        $itemIDs = $quote->getItemsCollection()->getAllIds();
+        $quoteTotalAmount = round($quote->getGrandTotal()*100);
+        $isCustomerGuest = $quote->getCustomerIsGuest();
+        $discountCode = str_replace(' ', '', $shippingAddress->getDiscountDescription());
+
+        $key = $quote->getId() . '_' . $isCustomerGuest . '_' . $discountCode . '_' .$quoteTotalAmount;
```

Comment:
I'm a little worried about using $quoteTotalAmount because in edge cases the customer could swap out items that are the same price but a different weight, in which case we would need to recalculate. What if we instead run through all skus and quantities and create a checksum? That way we are ensuring that the actual items are all accounted for?

Reply:
$quoteTotalAmount is also important still since the total could change based on other factors (customer group, shopping cart price rule that doesn't depend on coupon, etc). But I think my point was that we need to make sure that the items in the cart are the same.

Reply:
Disregard my last comments. Just saw that you're adding them at the end :)

Reply:
@daisy1754, Should I add product weight and qty there?

Reply:
I think weight is implied via sku. adding qty is probably a good idea.